lib/services/role.go

```diff
@@ -1558,6 +1592,147 @@ func (set RoleSet) CheckAccessToKubernetes(namespace string, kube *KubernetesClu
 }
 
+// CanImpersonateSomeone returns true if this checker has any impersonation rules
+func (set RoleSet) CanImpersonateSomeone() bool {
+	for _, role := range set {
+		cond := role.GetImpersonateConditions(Allow)
+		if !cond.IsEmpty() {
+			return true
+		}
+	}
+	return false
+}
+
+// CheckImpersonate returns nil if this role set can impersonate
+// a user and their roles, returns AccessDenied otherwise
+// CheckImpersonate checks whether current user is allowed to impersonate
+// users and roles
+func (set RoleSet) CheckImpersonate(currentUser, impersonateUser types.User, impersonateRoles []types.Role) error {
+	ctx := &impersonateContext{
+		user:            currentUser,
+		impersonateUser: impersonateUser,
+	}
+	whereParser, err := newImpersonateWhereParser(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// check deny: a single match on a deny rule prohibits access
+	for _, role := range set {
+		cond := role.GetImpersonateConditions(Deny)
+		matched, err := matchDenyImpersonateCondition(cond, impersonateUser, impersonateRoles)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if matched {
+			return trace.AccessDenied("access denied to '%s' to impersonate user '%s' and roles '%s'", currentUser.GetName(), impersonateUser.GetName(), roleNames(impersonateRoles))
+		}
+	}
+
+	// check allow: if matches, allow to impersonate
+	for _, role := range set {
+		cond := role.GetImpersonateConditions(Allow)
+		matched, err := matchAllowImpersonateCondition(ctx, whereParser, cond, impersonateUser, impersonateRoles)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if matched {
+			return nil
+		}
+	}
+
+	return trace.AccessDenied("access denied to '%s' to impersonate user '%s' and roles '%s'", currentUser.GetName(), impersonateUser.GetName(), roleNames(impersonateRoles))
+}
+
+func roleNames(roles []types.Role) string {
+	out := make([]string, len(roles))
+	for i := range roles {
+		out[i] = roles[i].GetName()
+	}
+	return strings.Join(out, ", ")
+}
+
+// matchAllowImpersonateCondition matches impersonate condition,
+// both user, role and where condition has to match
+func matchAllowImpersonateCondition(ctx *impersonateContext, whereParser predicate.Parser, cond types.ImpersonateConditions, impersonateUser types.User, impersonateRoles []types.Role) (bool, error) {
+	// an empty set matches nothing
+	if len(cond.Users) == 0 && len(cond.Roles) == 0 {
+		return false, nil
+	}
+	// should specify both roles and users, this condition is also verified on the role level
+	if len(cond.Users) == 0 || len(cond.Roles) == 0 {
+		return false, trace.BadParameter("the system does not support empty roles and users")
+	}
+
+	if !newMatcherFn(cond.Users)(impersonateUser.GetName()) {
+		return false, nil
+	}
+
+	roleMatch := newMatcherFn(cond.Roles)
+	for _, impersonateRole := range impersonateRoles {
+		if !roleMatch(impersonateRole.GetName()) {
+			return false, nil
+		}
+		ctx.impersonateRole = impersonateRole
```

Comment:
Am I understanding correctly that this basically sets `impersonateRole` inside the `ctx` that is in turn used inside `whereParser` which is created in CheckImpersonate above but is being used right below?

If I'm correct and there's no easy way to make this more explicit (like maybe create `whereParser` here instead of in the caller CheckImpersonate), I'd at least add a comment explaining this, took me a minute to understand